### PR TITLE
Enh/missing constraints

### LINKED ISSUE #149: What to do when no candidate is found in the rules generator
At least the 2 following cases are relevant:

1. Conflicts that cannot be resolved. Should those then be ignored ?
2. Dependencies missing in some requirements, e.g. package `A 1.0.0-1` depends on `B ^= 1.0.0`, `A 1.0.0-2` does not depend on B. For some reason, no candidate for `B ^= 1.0.0` can be found. It should be possible to install `A 1.0.0-2` (ignoring missing deps for `A 1.0.0-1`), or raise some kind of broken dependency exception if  `A 1.0.0-1` is requested.

### PR DESCRIPTION
**Pull request by johntyree**
This allows us to optionally ignore packages with missing dependencies or conflicts if they aren't tied directly to a job from the user.

If strict is `True`:
   - When a user specified job is broken, throw an exception.
   - When a dependency or conflict is broken log a WARNING.

If strict is `False`:
   - When a dependency or conflict is broken log an INFO.

Closes #149, #144

### CODE REVIEW

**Comment by johntyree** [comment]
Still need to check this on windows. I don't know of any broken packages of ours on Linux.

File: simplesat/rules_generator.py [hunk 1]
```diff
@@ -296,20 +310,42 @@ def _add_install_requires_rules(self, package, work_queue, requirements):
...
+                    # job requirement. Unfortunately, we don't track that
+                    # explicitly because we push all of the work through a
+                    # queue. As a proxy, we can examine the associated
+                    # requirements directly.
+                    if len(requirements) == 1:
+                        raise MissingInstallRequires(pkg_requirement, msg)
+                    else:
+                        logger.warning(msg)
+                else:
+                    logger.info(msg)
```
Comment (johntyree) [inline simplesat/rules_generator.py]: Missing a `return` statement.

Comment (johntyree) [inline simplesat/rules_generator.py]: [Nevermind](https://github.com/enthought/sat-solver/pull/156/files#diff-27965fdafe1fac1e643870b7f741db60R337), it's down below. We have to make our rule preventing installation first.

**Comment by johntyree** [comment]
Turns out `MiniSat` was not handling conflict discovery correctly during initialization. It never mattered before because we don't typically have totally broken package metadata. Now that we explicitly ban a package with broken metadata, we can end up with conflicts before we're even done reading all the rules. This fixes that.

**Comment by johntyree** [comment]
This should fix #144 entirely. I've confirmed that on master, it crashes or in some unfortunate cases, doesn't crash *or* solve the problem, while on this branch it does.

There are also tests that replicate the problem and verify that it is caught correctly now.

File: simplesat/rules_generator.py [hunk 2]
```diff
@@ -296,20 +310,43 @@ def _add_install_requires_rules(self, package, work_queue, requirements):
...
+                req_str = str(pkg_requirement)
+                msg = ("Blocking package {0!s}: no candidates found for"
+                       " dependency {1!r}").format(pkg_str, req_str)
+                if self.strict:
+                    # We only raise an exception if this comes directly from a
+                    # job requirement. Unfortunately, we don't track that
+                    # explicitly because we push all of the work through a
+                    # queue. As a proxy, we can examine the associated
+                    # requirements directly.
+                    if len(requirements) == 1:
```
Comment (cournape) [inline simplesat/rules_generator.py]: I don't understand the logic behind testing for `len(requirements) == 1` ?

**Comment by cournape** [comment]
The code LGTM, but I have one question regarding the logic

**Comment by johntyree** [comment]
I guess my long comment there didn't work, :disappointed:

The behavior differs depending on whether the broken package is directly
from a job requirement or just from a transitive dependency. Unfortunately,
the only way to know without a large refactor is to count the number of
requirements in the stack.

If there's only one it's a job, Otherwise it isn't.

**Comment by johntyree** [comment]
I made the comment a little more explicit. I'll merge after Travis gives the nod.

**Comment by cournape** [comment]
I understood the comment, but did not get `len(requirements) == 1` implied the corresponding requirement was a job requirement.

**Comment by cournape** [comment]
and I am not sure that I do yet :) I understand that for a job requirement, `len(requirements) == 1`, but not the opposite relationship

**Comment by johntyree** [comment]
Forcefully installed broken package discussion can move to https://github.com/enthought/sat-solver/issues/158.

Merging this.

**Pull request merged**